pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,25 +15,263 @@
+		klog.Errorf("Error when create collecting process: %v", err)
+		return
+	}
+	fa.collectingProcess = cp
+
+	ap, err := ipfix.NewIPFIXAggregationProcess(fa.collectingProcess.GetMsgChan(), 2, aggregationFields)
+	if err != nil {
+		klog.Errorf("Error when create aggreagation process: %v", err)
+		return
+	}
+	fa.aggregationProcess = ap
+
+	exportTicker := time.NewTicker(fa.exportInterval)
+	defer exportTicker.Stop()
+	go fa.collectingProcess.Start()
+	go fa.aggregationProcess.Start()
+	for {
+		select {
+		case <-stopCh:
+			fa.collectingProcess.Stop()
+			fa.aggregationProcess.Stop()
+			if fa.exportingProcess != nil {
+				fa.exportingProcess.CloseConnToCollector()
+			}
+			return
+		case <-exportTicker.C:
+			if fa.exportingProcess == nil {
+				err := fa.initExportingProcess()
+				if err != nil {
+					klog.Errorf("Error when initializing exporting process: %v", err)
+					// There could be other errors while initializing exporting process other than connecting to IPFIX collector,
+					// therefore closing the connection and resetting the process.
+					if fa.exportingProcess != nil {
+						fa.exportingProcess.CloseConnToCollector()
+						fa.exportingProcess = nil
+					}
+					continue
+				}
+			}
+			err := fa.aggregationProcess.ForAllRecordsDo(fa.sendFlowKeyRecords)
+			if err != nil {
+				klog.Errorf("Error when sending flow records: %v", err)
+				// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
+				// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
+				fa.exportingProcess.CloseConnToCollector()
+				fa.exportingProcess = nil
+				continue
+			}
+			for _, key := range fa.sentFlowKeys {
+				fa.aggregationProcess.DeleteFlowKeyFromMap(*key)
+			}
+			fa.sentFlowKeys = make([]*ipfixintermediate.FlowKey, 0)
+		}
+	}
+}
+
+func (fa *flowAggregator) sendFlowKeyRecords(key ipfixintermediate.FlowKey, records []ipfixentities.Record) error {
+	setSent := ipfix.NewSet(ipfixentities.Data, fa.templateID, false)
+	for _, record := range records {
+		elements := make([]*ipfixentities.InfoElementWithValue, 0)
+		for _, ieString := range fa.infoElementsOrder {
+			element, exist := record.GetInfoElementWithValue(ieString)
+			if !exist {
+				return fmt.Errorf("%s not present.", ieString)
+			}
+			elements = append(elements, element)
+		}
+		err := setSent.AddRecord(elements, fa.templateID)
+		if err != nil {
+			return fmt.Errorf("AddRecord failed, error: %v", err)
+		}
+	}
+	bytesSent, err := fa.exportingProcess.AddSetAndSendMsg(ipfixentities.Data, setSent.GetSet())
+	if err != nil {
+		return fmt.Errorf("Send data record for key %+v failed, error: %v", key, err)
+	}
+	fa.sentFlowKeys = append(fa.sentFlowKeys, &key)
+	klog.V(4).Infof("Sent key: %+v, sent bytes: %d, records num: %d\n", key, bytesSent, len(records))
+	return nil
+}
+
+func (fa *flowAggregator) sendTemplateRecord(templateSet ipfix.IPFIXSet) (int, error) {
```

Comment:
is there no utility function in `go-ipfix` to do that? @srikartati

Reply:
We provide an interface to send the filled-up template set (`AddSetAndSendMsg` or `SendSet` in go-ipfix 0.4.0), but the Set has to be filled by the user. Currently, we provide an interface (`AddRecord`) that takes in InfoElementWithValue objects and not name and enterprise ID.

However, in general, we can have an utility in go-ipfix in the next release that takes in element names and associated enterpriseID. The interface will build the template set and sends it.